### PR TITLE
feat: add FDv2 protocol foundation types and state machine

### PR DESCRIPTION
**Pull request by kinyoklion**
Adds the base FDv2 protocol types and the protocol handler.

---

> [!NOTE]
> **Medium Risk**
> Introduces new core protocol parsing/state-machine code for FDv2 payload handling; while mostly additive, bugs here could affect correctness of future streaming/polling update ingestion.
> 
> **Overview**
> Adds the foundational FDv2 client protocol layer: new wire-model types (`FDv2Event`, `ServerIntentData`, etc.), a `Payload`/`Update` representation with `Selector`, and a pure `FDv2ProtocolHandler` state machine that turns incoming events into `ProtocolAction`s (emit payloads, surface protocol/server errors, handle `goodbye`, ignore unknown kinds/heartbeat).
> 
> Adds a `flag-eval` mapper (`mapUpdatesToItemDescriptors`) to convert FDv2 updates into flag-store `ItemDescriptor`s (including tombstones on delete and rethrowing on unparseable evaluations), plus comprehensive unit tests covering parsing, state transitions, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 216e015ee3844af097987c5fcfd52b36659581dd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kinyoklion** [comment]
bugbot review

**Comment by kinyoklion** [comment]
bugbot review

File: packages/common_client/lib/src/data_sources/fdv2/selector.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  /// The opaque state string from the server.
+  final String state;
+
+  /// The payload version associated with this selector.
+  final int version;
+
+  const Selector({required this.state, required this.version});
+
+  bool get isEmpty => state.isEmpty;
+  bool get isNotEmpty => state.isNotEmpty;
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/selector.dart]: If a selector has a non-zero version, but empty state, it will statisfy `mySelector.isEmpty == true` but will not satisfy `mySelector == Selector.empty`.

File: packages/common_client/lib/src/data_sources/fdv2/source_result.dart [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  final DateTime? freshness;
+
+  /// Whether the data should be persisted to the cache.
+  final bool persist;
+
+  const ChangeSetResult({
+    required this.payload,
+    this.environmentId,
+    this.freshness,
+    this.persist = true,
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/source_result.dart]: Consider not having a default to have the source make the decision.

File: packages/common_client/lib/src/data_sources/fdv2/flag_eval_mapper.dart [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    if (update.kind != flagEvalKind) {
+      continue;
+    }
+
+    if (update.deleted) {
+      result[update.key] = ItemDescriptor(version: update.version);
+    } else if (update.object != null) {
+      try {
+        final evalResult =
+            LDEvaluationResultSerialization.fromJson(update.object!);
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/flag_eval_mapper.dart]: `} else if (update.object case final object?) {` can get rid of the `!` based on some AI digging.

File: packages/common_client/lib/src/data_sources/fdv2/protocol_handler.dart [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        _tempType = PayloadType.partial;
+        return _actionNone;
+      case IntentCode.none:
+        _state = ProtocolState.changes;
+        _tempUpdates = [];
+        _tempType = PayloadType.partial;
+        return _processIntentNone(payload);
+      case null:
+        _logger.warn('Unable to process intent code '
+            "'${payload.intentCode}'.");
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/protocol_handler.dart]: This will log `Unable to process intent code 'null'.` which may be confusing.

File: packages/common_client/lib/src/data_sources/fdv2/protocol_types.dart [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+  const FDv2EventsCollection({required this.events});
+
+  factory FDv2EventsCollection.fromJson(Map<String, dynamic> json) {
+    final eventsList = json['events'] as List<dynamic>?;
+    return FDv2EventsCollection(
+      events: eventsList
+              ?.map((e) => FDv2Event.fromJson(e as Map<String, dynamic>))
+              .toList() ??
+          [],
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/protocol_types.dart]: That is some wonky formatting.

File: packages/common_client/lib/src/data_sources/fdv2/payload.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  });
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Update &&
+          kind == other.kind &&
+          key == other.key &&
+          version == other.version &&
+          deleted == other.deleted;
```
Comment (tanderson-ld) [inline packages/common_client/lib/src/data_sources/fdv2/payload.dart]: Could this not including object as part of the comparison cause confusion?  Maybe an explicit comparison function for just the top level data could provide clarity.

File: packages/common_client/lib/src/data_sources/fdv2/selector.dart (hunk 1 above)
Comment (kinyoklion) [reply packages/common_client/lib/src/data_sources/fdv2/selector.dart]: I can adjust, but empty isn't a valid state, so I think that is still empty. Or potentially invalid.

Comment (kinyoklion) [reply packages/common_client/lib/src/data_sources/fdv2/selector.dart]: I'll just adjust empty for now. I want it to be a consistent constraint.

File: packages/common_client/lib/src/data_sources/fdv2/payload.dart (hunk 6 above)
Comment (kinyoklion) [reply packages/common_client/lib/src/data_sources/fdv2/payload.dart]: This is missing object, and should have it.

**Review by tanderson-ld — APPROVED** [review]
